### ISSUE TITLE
Different View of Avalon: Task-Subset-Families

### ISSUE THREAD

**Issue by davidlatwe** [getavalon/core#294]
## Proposal

Get `Task` involved, and subdivide the concept of `family`, from *`family is subset; a type of data`* to *`family is one feature of subset; a type of data portion`*, like :
```
ModelDraft = ["pipeline.mesh", "pipeline.transform"]
ModelRender = ["pipeline.mesh", "pipeline.transform", "pipeline.uv"]
```
Maybe this will bring some possible usage and benefit from using `families` in subset.

#### Another way to look at

* Task
Define what baseType of subset you want to process in this session.
e.g., Model, Rig, Look..

* Subset
Defined by one or more families, a Task related creation or variation, a component of asset.
e.g., ModelDraft, ModelBlend, ModelDeform...

* Family
A type of subdata, a component or feature of a subset.
e.g., mesh, transform, uv, controller, joint...

## Implementation

Avalon needs to support `families` in CreatorPlugin, for further digging.
Here's what I have now. eb6d6e5

## Pros

* More bound with `Task`, like using `Task` name for subset searching.
* Can Implicit family/families to artists.
* When publishing, with more specific family type, no validation plugin needs to be toggled.

Simply put, take away some mouse clicks or keystrokes from artist, decrease the human error. Maybe there are more benefit than listed above.

## Cons

* Subset-families needs to be predefined ( how many variation == how many Creator ).
* May take a lot change on Creator and Loader Tool for support.

Currently having some result on my new Creator Tool ( modified from original ) , but not sure Loader Tool needs how much modification to support this.

Things might get more complex. Need test to find more.

---

What do you think ? :)

**Comment by davidlatwe** [comment]
Note: This doesn't meant to replace current workflow or lose the backward compatibility, but adding more support for another kind of pipeline style built with Avalon.

**Comment by mottosso** [comment]
Thanks @davidlatwe. I'm having a bit of trouble understanding what this is. I think it would help with an example, a description of the problem this solves along with what motivated you to come up with it. Could you write a little bit about that?

**Comment by davidlatwe** [comment]
Yes, of course.

Currently the variant of subset is defined with creator tools, by artist selecting family and choosing variant type ( subset ) pre-existed in database or defining a new one on the fly.

Later in publish, artist may toggle on/off some validation plugin for some type of subset variation.

So I was thinking that if I want to let artist to have less button to press, I have to predefined more family type like `studio.modelLOD`, `studio.rigController` or `studio.modelNoUV`... . This is okay, of course.

However, I wonder what if not to use family like that, but use something more divided, like `studio.transform`, `studio.mesh`, `studio.uv`, `studio.animCurve`, `studio.joint`..., and use them to assemble the subset I want. Then there comes the subset with families.

And the Task, I found that if we separate the concept of subset and family, we can bring in Task as the base of subset, e.g., `Model` Task will only produce Model type of subsets like `modelDraft`, `modelStatic`, `modelDeform`.

This is what I was thinking, like another way to build pipeline with Avalon.

But the key part is Creator have to support `families`, not sure this is heading to good or bad.

**Comment by mottosso** [comment]
Ah, ok. I think I understand.

Would it be possible if you could select multiple families in the Creator? For example, rather than having a `rig` and `model`, you had your granular families, like `mesh` and `transform` and then the artist could simply pick the ones he wanted?

Overall I think this workflow is subpar and that it'll bite you later on. One of the strengths of *not* doing it this way is that you get to define what assets flow through your pipeline and how they look. Once you know what a specific asset looks like, like a `simRig`, then you can build tools that operate on `simRig` and perhaps how they interact with a `deformationRig`, `furRig` or even other types of assets like `collisionGeometry` or `skeleton`.

If you instead delegate the low-level details to your artist, you leave little room for yourself to continue building upon what they produce.

An ideal pipeline to me is one with very very few different kinds of assets, where each asset has lots of tooling, and can be used in lots of creative ways. The alternative, and opposite, is having no two assets be alike, and therefore little or no tools that know how to work with them.

Having said that, I'm making a lot of assumptions about what you are looking to do based on the little information I had to go on, it's possible it doesn't apply to you actual goals.

**Comment by davidlatwe** [comment]
> Would it be possible if you could select multiple families in the Creator? For example, rather than having a rig and model, you had your granular families, like mesh and transform and then the artist could simply pick the ones he wanted?

Ah, I missed a spot in last comment. No, in my Creator will only shows predefined subset name, the family/families are all Implicit from artist.
And in this workflow, all family/families are like stay in under ground, and explicit subset name.

> Once you know what a specific asset looks like, like a simRig, then you can build tools that operate on simRig and perhaps how they interact with a deformationRig, furRig or even other types of assets like collisionGeometry or skeleton.

I still see those asset ( subset ) by the `families` data they kept, so I think I won't lose how they looks like. 

The granular families, this might mainly for publish.
In my imaging, no matter how two type of subset be alike, there must be some different in them, and can take the different out as a family to identify. 

Therefore, I would have many CreatorPlugin, each of them have different families list and subset name. In publish, some plugins could be widely used, and some of them is very specific.

> Having said that, I'm making a lot of assumptions about what you are looking to do based on the little information I had to go on, it's possible it doesn't apply to you actual goals.

Sorry about that, this is still a rough idea.

**Comment by mottosso** [comment]
Don't be sorry! I just meant that I might not be fully understanding the idea just yet and that because of it my thoughts of it might not apply.

> Ah, I missed a spot in last comment. No, in my Creator will only shows predefined subset name, the family/families are all Implicit from artist. And in this workflow, all family/families are like stay in under ground, and explicit subset name.

Ok, I think you'll need to come up with an example here. It sounds like the artists cannot type in a name, but has a set of pre-defined names to choose from, like `studio.uv` and `studio.transform` and that the name of the task is somehow appended to this to form a final subset (and family?) name.

If you could take me through the steps I think that'll help.

For example..

1. Artist opens Maya
2. Artist makes a character model
3. Artist opens up Creator
4. Artist selects ...
5. Artist types ...
6. Artist clicks "Create"
7. Artist publishes

Better yet, it sounds like it's something you could build and test with a small GUI. That would really bring the point home I think.

**Comment by BigRoy** [comment]
Just wanted to chime in and say @davidlatwe thanks for spawning some extra ideas. Like @mottosso I'm currently having a hard time understanding exactly the workflow you're describing.

What I get from your _pros_ and _cons_ you want to allow the artist to pick the "subset" which for you is the extra "version" like *Draft* or *Render* which would then differ in what it ends up checking. If this is what you're looking for you could even create more families currently (like you're describing with adding more Creators), they could potentially be family `modelDraft` and `modelRender`, then users can create any amount of `modelRender` publishes within an asset, like `modelRenderDefault`, `modelRenderDamaged` - where "Default" and "Damaged" are the subsets.

But then again, I might not be fully understanding what you're going for so I'd be all ears for an example like Marcus mentioned.

**Comment by davidlatwe** [comment]
Thanks, @mottosso .

>  It sounds like the artists cannot type in a name, but has a set of pre-defined names to choose from

Yes. But they only choose the `subset`.
Here is my current modified Creator GUI

![](https://i.imgur.com/UeIxejK.png)

I also locked the `Asset` name input, left only subset type for artist to pick. ( will have some kind of mark to notify which subset type has pre-existed in database )
And the families are hidden in CreatorPlugin, and it's fixed and different from each Creator.
```python
class ModelDeformCreator(avalon.maya.Creator):
    """Polygonal geometry for animation"""

    label = name = "modelDeform"
    families = ["studio.transform",
                "studio.mesh", 
                "studio.faceCountLimit", 
                "studio.uv", "studio.uvTile",
                "studio.uuid"]
```

---

I think I can drop this "Task-Subset-Families" thing for now, and back to the core of my problem.

My main issue want to solve are :
1) Let artist to deal only with `subset name` and hide the `family` from them.
2) Let artist to keep away from toggle validation plugins in publish.

for 1), I think it could easy solve by my customized GUI and predefined more CreatorPlugins with different `name`, but same `family`. e.g.
```python
class ModelDraftCreator(avalon.maya.Creator):
    label = name = "modelDraft"
    family = "studio.modelDraft"

class ModelDeformCreator(avalon.maya.Creator):
    label = name = "modelDeform"
    family = "studio.modelDeform"
```
for 2), that's why I came up with this issue, to divide more family from the difference between subsets that are alike, for some validation plugin can be more specific on each of them.
But, it's rough now and letting `subset` support `families` might have bad effect with other things, like easily confused with current concept ( needs more polish on this :S ).

So, for 2), maybe I could store an info about which validation plugin need to on/off during publish inside my CreatorPlugin, or I can memorize them into the subset data after publish ?

**Comment by davidlatwe** [comment]
Just found out that @BigRoy came in while I was typing.
Thanks, please let me know if there are anything needs to be point out or something is over thinking.

**Comment by davidlatwe** [comment]
Oh, there is another reason for this "subset-families" thing, I was thinking the possibility to use "Multiple families workflow" from Pyblish to here in Maya with Avalon recently, so just thought this might be it.

**Comment by mottosso** [comment]
Ok, that's great. But what I'd really like to see is an example. Who would use this? When? Under what circumstance? What problem does it solve?

You mention you can create multiple families to resolve this, and that this is an alternative to that. But I'm still not seeing why you need an alternative. Why not just make more families?

Try and think of an example circumstance, perhaps involving one wanting to create and publish an asset, and what problems the artist may run into without this feature.

**Comment by davidlatwe** [comment]
> Why not just make more families?

Oh oh, Yes, I could. I mentioned that in this comment ( maybe you missed this one ? ) https://github.com/getavalon/core/issues/294#issuecomment-340370138
Here :
> for 1), I think it could easy solve by my customized GUI and predefined more CreatorPlugins with different `name`, ~but same~ and different `family`. e.g.

Next
> Who would use this? When? Under what circumstance? What problem does it solve?

For the reason why need this alternative, it's from here:
> 2) Let artist to keep away from toggle validation plugins in publish.

Yes, I do need to elaborate more about this. I will head back with an example later today !

**Comment by davidlatwe** [comment]
Say we have following kinds of task :
* **draft model** for a character asset, mesh will be simple and `does not matter it has UV or not`.
* **deform model** for a character asset, this will be a high poly model for render and `need to have UV`.
* **deform LOD model** for crowd asset, will be `a few proxy, low model` for render and `need to have UV`.
* **blendShape model** for a character asset, `need the same UV` as original mesh, `transform can not freeze`.

Something like above, each of them would have different validation condition, like `polyCount`, `uv`, `transform`, `uuid`.

Now the validate plugins I have for example and assign family to them :
* mesh normal : `modelDraft`, `modelDeform`, `modelDeformLOD`, `modelBlendShape`
* mesh no-manifold : `modelDraft`, `modelDeform`, `modelDeformLOD`, `modelBlendShape`
* mesh double vertices : `modelDraft`, `modelDeform`, `modelDeformLOD`, `modelBlendShape`
* mesh poly count : `modelDeformLOD`
* mesh has uv : `modelDeform`, `modelDeformLOD`, `modelBlendShape`
* transform freeze : `modelDraft`, `modelDeform`, `modelDeformLOD`
* uuid check : `modelDraft`, `modelDeform`, `modelDeformLOD`

And each Creator has each family above. ( `modelDraft`, `modelDeform`, `modelDeformLOD`, `modelBlendShape` )

I can work with this way without error, also let artist easier.
But this is the point during my thinking, where I came up with this issue.
**Not because I got problem, but another way to organize them.**
( Maybe my expression missed lead about this )

---

So, another way to organize, in the example above, I have assigned many families in each validation plugins, how about grouping them with family like this :

* mesh normal : `mesh`
* mesh no-manifold : `mesh`
* mesh double vertices : `mesh`
* mesh poly count : `faceLimit`
* mesh has uv : `hasUV`
* transform freeze : `freeze`
* uuid check : `uuid`

And for each Creator :
```python
class ModelDraftCreator(avalon.maya.Creator):
    label = name = "modelDraft"
    families = ["mesh", "freeze", "uuid"]

class ModelDeformCreator(avalon.maya.Creator):
    label = name = "modelDeform"
    families = ["mesh", "hasUV", "freeze", "uuid"]

class ModelDeformLODCreator(avalon.maya.Creator):
    label = name = "modelDeformLOD"
    families = ["mesh", "hasUV", "freeze", "faceLimit", "uuid"]

class ModelBlendShapeCreator(avalon.maya.Creator):
    label = name = "modelBlendShape"
    families = ["mesh", "hasUV"]
```
And that's it, move some family weight from validation plugin to creator.

I think this is the example to express what I meant to.
Does this help ?

**Comment by davidlatwe** [comment]
Ok, after some more practice on my end, yap, this granular families cost me more energy then I thought to work with.

I end up with this :

In my config, I have a `family.py`, stores all the family I need, and some families preset, to help me organize.
```python
# family.py
class Pipeline:
    # Model
    modelDraft = 'reveries.modelDraft'
    modelAnima = 'reveries.modelAnima'
    modelDeform = 'reveries.modelDeform'
    modelStatic = 'reveries.modelStatic'
    modelDeformLOD = 'reveries.modelDeformLOD'
    modelStaticLOD = 'reveries.modelStaticLOD'


class Presets:
    # basic model family
    model = [
        Pipeline.modelDraft,
        Pipeline.modelAnima,
        Pipeline.modelDeform,
        Pipeline.modelStatic,
        Pipeline.modelDeformLOD,
        Pipeline.modelStaticLOD
    ]
    # these need to check the poly count
    polyLimit = [
        Pipeline.modelDeformLOD,
        Pipeline.modelStaticLOD
    ]
```

Then in my publish plugin I can call the families preset I need to use.
```python
class ValidateMeshNormal(pyblish.api.InstancePlugin):
   families = family.Presets.model

class ValidatePolyFaceCount(pyblish.api.InstancePlugin):
   families = family.Presets.polyLimit
```

Did see making subset to have `families` will introduce some extra work, but did not know it's a lot.

---

And the `Task` part of this issue, was thinking there has some overlap between `Subset` and `Task`, so maybe can use `Task` to help defining or searching `Subset`.

---

Phew... Thanks for you guys' patient ! :sweat_smile:

**Comment by mottosso** [comment]
Don't be sorry, this was an excellent exercise. :) Keep your ideas coming!

**Comment by davidlatwe** [comment]
Hi, still a bit confuse about how Avalon using Pyblish. :sweat_smile: 

The following things I want to ask is alike but not exactly the same with this issue, but lead to the same point inside of it, so I think this is a fair place to go on.

---

I now have a type of rig for crowd sim, and comparing to normal rig, that rig for crowd agent has a few extra validations need to be run. ( like checking there isn't any joint overlapping each other in same position, it's okay or needed for normal character rig, but not good for crowd sim )

To do this, I have a few workaround :

* option 1), I could add one more family named `rigCrowd` into my publish plugins to make things work.

* Or 2), I could use family `rig` to that subset and make those extra validation support family `rig` but being optional or just sending warning for also being able to exist in normal rig publishing.

* 3), during publish, I can use Collectors to check instance's subset-type, and do `instance.data["families"] = ["rig", "rig.crowd"]` if the subset-type is `rigCrowd`, and upload the subset to data-base *without* `data["families"]`, just keeping it's `family = "rig"`.

Those workaround above all works for me, and option 3) is the way I like the most, because it's the one which closest to my perspective of the relationship between subset and family.

But still curious, if we can publish instance with `families` in Pyblish, why Avalon only see a subset == a family ?

Isn't that the subset like `rigCrowd` is actually a subset which based on family `rig` and plus one extra family `rig.crowd` (a variation of family `rig`) more logical for running validation in Pyblish ?
Why not making subset to be able to use multiple family (`families`) ? Using `families` to describe the variation of `family` (subset).

Not that I keep trying to involve multiple family and coming up with this, just happen to got in to another scenario and wondering, are there any reason for not doing this ?

Thanks ! :)

**Comment by mottosso** [comment]
Could you make it two different families, both of which are supported by relevant plug-ins?

```
plugin1.families = [rig]
plugin2.families = [rig, crowdRig]
plugin3.families = [crowdRig]
```

Then `rig` would be validated by `plugin1` and `plugin2`, whereas `crowdRig` would be validated by `plugin2` and `plugin3`, and they would both be validated by `plugin2`.

I think the way to think about the family of a subset is that of a type of asset. It's either a rig for animation, or a rig for crowds. If they share validations, then this would be one way of achieving that.

**Comment by davidlatwe** [comment]
> Could you make it two different families, both of which are supported by relevant plug-ins?

That's currently how I did.

> I think the way to think about the family of a subset is that of a type of asset. 

Yes... this is simpler to work with. :)
But, I don't understand, if Avalon is not going to make subset with families, at least for now, then why the version data stored in database use `families` ? Is it for "even use `families` in the future, won't affect data" ?

**Comment by mottosso** [comment]
The single family is stored because they are later used when loading the published asset. The reason it's `families` as opposed to `family` is, like you said, because in the future it might be interesting to support multiple families. Using the plural of the word simply means that door is left open.

Most Pyblish instances that I work with have a single family, and in fact the original keyword was `family` (and is still supported) but today I use `families` even I only have one, simply because I wouldn't have to change the underlying schema if/when I find that I need more than one.

On top of that, I find it easier to work with having the same word used in both plug-ins and instances. Especially since they are so related.

**Comment by davidlatwe** [comment]
Thank you @mottosso , really appreciate this, really make sense.

I am really interested in the possibility between family and subset, like, maybe a subset could be describe as a family type of data in some certain task usage, or something else, and what would that affect the way to manage and the way of Tools' UI design.

Maybe you guys have already thought about these, I am not sure. But just like you commented in other issue about the "Use of Avalon", I am interesting in that idea. And since there is already have a solid way to run the pipeline, and the design of Avalon is really giving a certain amount of freedom, I will give some of my thought a shot in my config to do some experiment.

> Most Pyblish instances that I work with have a single family, and in fact the original keyword was family (and is still supported) but today I use families even I only have one, simply because I wouldn't have to change the underlying schema if/when I find that I need more than one.
>
>On top of that, I find it easier to work with having the same word used in both plug-ins and instances. Especially since they are so related.

Thanks again, it's really helpful.

**Comment by mottosso** [comment]
> Edit : Actually, allow me to take those back, maybe I should keep things simple as is, sorry about that.

Ok. :) I was otherwise looking forward to the results of your experiment. :)

> I will give some of my thought a shot in my config to do some experiment.

**Comment by davidlatwe** [comment]
Ooh, those idea still ringing in my head and hard to go off, but it's so rough now and was thinking that I should finish my `avalon config` debt first ( soon I guess ), before I say I want to do the crazy stuff. :P

Anyway, if luckily having good things work out, definitely share them.
